driver/src/metrics/http_metrics.rs

```diff
@@ -0,0 +1,197 @@
+    }
+}
+
+/// A common trait shared between labeled and unlabeled subsystems for getting
+/// the label values to use for metrics.
+pub trait LabelValues {
+    fn label_values(&self) -> [&'static str; 2];
+    fn initialize_histogram_labels(histograms: &HistogramVec);
+}
+
+/// A trait abstracting a labeled subsystem, that is a specific use of an HTTP
+/// client where the HTTP requests should be differenciated.
+pub trait LabeledSubsystem: LabelValues + Sized {
+    fn name() -> &'static str;
+    fn all_labels() -> &'static [Self];
+    fn label(&self) -> &'static str;
+}
+
+/// A trait abstracting a labeled subsystem, that is a specific use of an HTTP
+/// client where the HTTP requests should not be differenciated.
+pub trait UnlabeledSubsystem: Default + LabelValues {
+    fn name() -> &'static str;
+}
+
+macro_rules! subsystem {
+    (
+        $(#[$attr:meta])*
+        pub enum $name:ident as $n:tt {
+            $($variant:ident => $label:tt,)*
+        }
+    ) => {
+        $(#[$attr])*
+        #[derive(Clone, Copy, Debug, Eq, PartialEq)]
+        pub enum $name {
+            $(
+                $variant,
+            )*
+        }
+
+        impl LabelValues for $name {
+            fn label_values(&self) -> [&'static str; 2] {
+                [Self::name(), self.label()]
+            }
+
+            fn initialize_histogram_labels(histograms: &HistogramVec) {
+                for label in Self::all_labels() {
+                    histograms.with_label_values(&label.label_values());
+                }
+            }
+        }
+
+        impl LabeledSubsystem for $name {
+            fn name() -> &'static str {
+                $n
+            }
+
+            fn all_labels() -> &'static [Self] {
+                const ALL: &[$name] = &[$($name::$variant),*];
+                ALL
+            }
+
+            fn label(&self) -> &'static str {
+                match self {
+                    $(
+                        $name::$variant => $label,
+                    )*
+                }
+            }
+        }
+    };
+
+    (
+        $(#[$attr:meta])*
+        pub struct $name:ident as $n:tt;
+    ) => {
+        $(#[$attr])*
+        #[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
+        pub struct $name;
+
+        impl LabelValues for $name {
+            fn label_values(&self) -> [&'static str; 2] {
+                [Self::name(), ""]
+            }
+
+            fn initialize_histogram_labels(histograms: &HistogramVec) {
+                histograms.with_label_values(&Self::default().label_values());
+            }
+        }
+
+        impl UnlabeledSubsystem for $name {
+            fn name() -> &'static str {
+                $n
+            }
+        }
+    };
+}
+
+subsystem! {
+    /// A label for `web3` HTTP requests. This subsystem is used by the HTTP
+    /// transport implementation.
+    pub enum Web3Label as "web3" {
+        Call => "call",
+        EstimateGas => "estimate_gas",
```

Comment:
Should sendTx be also a specific type?

Reply:
Don't think so, it just adds the tx to the mem-pool and is usually pretty fast and returns just a tx hash. If we were also tracking post body sizes, then yes we could definitely add it.